### PR TITLE
[Chinese text normalization]PR for chinese TN part in text_normalization

### PR DESCRIPTION
**Pull request by mzxcpp**
# What does this PR do ?

new PR for previous chinese TN

**Collection**: [NeMo/norm_text_processing/text_normalization]
I think have covered the problems mentioned previously.And I make a new pr to make it clear.

### CODE REVIEW

**Comment by dophist** [comment]
@BuyuanCui @ekmb  This PR continues https://github.com/NVIDIA/NeMo/issues/4543 & https://github.com/NVIDIA/NeMo/pull/4638 , ready for another round of reviews.

File: nemo_text_processing/text_normalization/zh/data/char/lower_to_upper.tsv [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+q	Q
+r	R
+s	S
+t	T
+u	U
+v	V
+w	W
+x	X
+y	Y
+z	Z
```
Comment (XuesongYang) [inline nemo_text_processing/text_normalization/zh/data/char/lower_to_upper.tsv:26]: Thanks for your PR. Is there any reason we need this mapping? It seems a trivia op in Python.

File: nemo_text_processing/text_normalization/zh/data/char/upper_to_lower.tsv [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+Q	q
+R	r
+S	s
+T	t
+U	u
+V	v
+W	w
+X	x
+Y	y
+Z	z
```
Comment (XuesongYang) [inline nemo_text_processing/text_normalization/zh/data/char/upper_to_lower.tsv:26]: Same question as the above. Is there any reason we need this mapping as a standalone file? There are lower() and upper() functions in Python.

File: nemo_text_processing/text_normalization/zh/data/money/currency.tsv [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+AUD	澳元
+A$	澳元
+SUR	卢布
+₽	卢布
+DEM	马克
+DM	马克
+FRF	法郎
+CAD	加元
+CAD$	加元
+FRF	法郎
```
Comment (XuesongYang) [inline nemo_text_processing/text_normalization/zh/data/money/currency.tsv:21]: nit: is it possible to store currency codes (e.g. USD) and symbols (e.g. $) in separate files?

Comment (XuesongYang) [reply nemo_text_processing/text_normalization/zh/data/money/currency.tsv:21]: It seems `currency.tsv` file only consider a bunch of currencies. My guess is that the selected currency set occurs most frequently in the Chinese datasets used here. Could you please share any thoughts? Given https://www.eurochange.co.uk/travel-money/world-currency-abbreviations-symbols-and-codes-travel-money, it seems many more currencies worth to add in this file.

File: nemo_text_processing/text_normalization/zh/verbalizers/time.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+class Time(GraphFst):
+    '''
+        tokens { time { h: "1" m: "02" s: "36" } } -> 一点零二分三十六秒
+        tokens { time { suffix "am"  hours: "1" minutes: "02" seconds: "36" } } -> 上午一点零二分三十六秒
+    '''
+
+    def __init__(self, deterministic: bool = True, lm: bool = False):
+        super().__init__(name="time", kind="verbalize", deterministic=deterministic)
+        graph_digit = pynini.string_file(get_abs_path("data/cardinal/digit.tsv"))
+        graph_ten = pynini.string_file(get_abs_path("data/cardinal/digit_teen.tsv"))
```
Comment (XuesongYang) [inline nemo_text_processing/text_normalization/zh/verbalizers/time.py:36]: I observed there are three places loaded `"data/cardinal/digit_teen.tsv"`, but they declared two different variables below, please unify the name space. 
* `graph_ten`: line 36 in `text_normalization/zh/verbalizers/time.py`; line 30 in `text_normalization/zh/verbalizers/date.py`.
* `graph_teen`: line 48 in `text_normalization/zh/taggers/cardinal.py`.

File: nemo_text_processing/text_normalization/zh/data/char/lower_to_upper.tsv (hunk 1 above)
Comment (mzxcpp) [reply nemo_text_processing/text_normalization/zh/data/char/lower_to_upper.tsv:26]: In this case,we want to process in graph and it could be set as a part of postprocessor instead of applying to string in public normalize.py.I have the plan to add chinese upper and lower forms in this tsv when I got engough tsv data.just like trans "壹“ to ”一“

File: nemo_text_processing/text_normalization/zh/taggers/tokenize_and_classify.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            char = Char(deterministic=deterministic)
+            fraction = Fraction(deterministic=deterministic)
+            math_symbol = MathSymbol(deterministic=deterministic)
+            money = Money(deterministic=deterministic)
+            measure = Measure(deterministic=deterministic)
+            time = Time(deterministic=deterministic)
+            whitelist = Whitelist(deterministic=deterministic)
+
+            classify = pynini.union(
+                pynutil.add_weight(date.fst, 0.4),
```
Comment (yzhang123) [inline nemo_text_processing/text_normalization/zh/taggers/tokenize_and_classify.py:77]: could you please keep class weights between [1.0, 1.1] and default chars 100, @fayejf, Alex Cui for help

File: tests/nemo_text_processing/zh/data_text_normalization/test_cases_time.txt [hunk 6]
```diff
@@ -0,0 +1,4 @@
+8月16号12:00之前 -> 八月十六号十二点之前
```
Comment (yzhang123) [inline tests/nemo_text_processing/zh/data_text_normalization/test_cases_time.txt:1]: ~ as delimiter

File: tests/nemo_text_processing/zh/test_char.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+
+class TestChar:
+    normalizer_zh = Normalizer(lang='zh', cache_dir=CACHE_DIR, overwrite_cache=False)
+
+    @parameterized.expand(parse_test_case_file('zh/data_text_normalization/test_cases_char.txt'))
+    @pytest.mark.run_only_on('CPU')
+    @pytest.mark.unit
+    def test_norm_money(self, expected, test_input):
+        preds = self.normalizer_zh.normalize(test_input, n_tagged=50)
```
Comment (yzhang123) [inline tests/nemo_text_processing/zh/test_char.py:29]: change `preds= self.normalizer_zh.normalize(test_input, n_tagged=50) ` to pred= self.normalizer_zh.normalize(test_input,)` this is deterministic normalization

and assert expected in preds -> assert expected ==  pred

File: tests/nemo_text_processing/zh/test_char.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+from ..utils import CACHE_DIR, parse_test_case_file
+
+
+class TestChar:
+    normalizer_zh = Normalizer(lang='zh', cache_dir=CACHE_DIR, overwrite_cache=False)
+
+    @parameterized.expand(parse_test_case_file('zh/data_text_normalization/test_cases_char.txt'))
+    @pytest.mark.run_only_on('CPU')
+    @pytest.mark.unit
+    def test_norm_money(self, expected, test_input):
```
Comment (yzhang123) [inline tests/nemo_text_processing/zh/test_char.py:28]: fix name

File: tests/nemo_text_processing/zh/test_sparrowhawk_normalization.sh [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+testTNMoneyText() {
+  input=$PROJECT_DIR/zh/data_text_normalization/test_cases_money_text.txt
+  runtest $input
+}
+testTNCharText() {
+  input=$PROJECT_DIR/zh/data_text_normalization/test_cases_char_text.txt
+  runtest $input
+}
+testTNClockText() {
```
Comment (yzhang123) [inline tests/nemo_text_processing/zh/test_sparrowhawk_normalization.sh:32]: change to time

File: tests/nemo_text_processing/zh/test_time.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+from ..utils import CACHE_DIR, parse_test_case_file
+
+
+class TestClock:
+    normalizer_zh = Normalizer(lang='zh', cache_dir=CACHE_DIR, overwrite_cache=False)
+
+    @parameterized.expand(parse_test_case_file('zh/data_text_normalization/test_cases_time.txt'))
+    @pytest.mark.run_only_on('CPU')
+    @pytest.mark.unit
+    def test_norm_clock(self, expected, test_input):
```
Comment (yzhang123) [inline tests/nemo_text_processing/zh/test_time.py:28]: change to time

File: tests/nemo_text_processing/zh/test_time.py [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+# limitations under the License.
+
+import pytest
+from nemo_text_processing.text_normalization.normalize import Normalizer
+from parameterized import parameterized
+
+from ..utils import CACHE_DIR, parse_test_case_file
+
+
+class TestClock:
```
Comment (yzhang123) [inline tests/nemo_text_processing/zh/test_time.py:22]: change to time

**Review by yzhang123 — CHANGES_REQUESTED** [review]
Thank you so much! 
A few comments: 
1. could you please add more test cases for every class? i noticed there are no cardinal tests, decimals tests etc.
2. Also our pipeline does not run our code, so could you please make sure the pytests run? you do this by cd NeMo/tests; pytest nemo_text_processing/zh/test_*.py --cpu --tn_cache_dir cache, only 2 out of all tests passed
3. also make sure the sparrowhawk tests are passing, by running cd NeMo/tools/text_processing_deployment; bash export_grammars.sh --GRAMMARS=tn_grammars  --LANGUAGE=zh --MODE=test --OVERWRITE_CACHE=True
4. could you adjust the weights of each semiotic class to be between [1, 1.01] see https://github.com/NVIDIA/NeMo/blob/main/nemo_text_processing/text_normalization/en/taggers/tokenize_and_classify.py#L165, 

The rest is ok, for language specific details like coverage of test cases, @fayejf and AlexCui could you add your review?

File: nemo_text_processing/text_normalization/zh/README.md [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+### 2.3 Post-Processing
+#### Punctuation Removal
+If enabled, punctuations are removed.
+
+#### Uppercase or Lowercase Conversion
+If enabled, English letters are converted to uppercases / lowercases
+
+#### Out-Of-Vocabulary(OOV) Tagger
+If enabled, OOV chars are tagged with `<oov>` and `</oov>`, e.g.:
```
Comment (yzhang123) [inline nemo_text_processing/text_normalization/zh/README.md:123]: why do we need this? can't we pass through everything as plain text? this is at most preprocessor,  but i would keep this optional and let the user set it

**Review by yzhang123 — CHANGES_REQUESTED** [review]

File: nemo_text_processing/text_normalization/zh/taggers/cardinal.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+        5       -> cardinal{ integer: "五"}
+        12      -> cardinal{ integer: "十二"}
+        213     -> cardinal{ integer: "二百一十三"}
+        3123    -> cardinal{ integer: "三千一百二十三"}
+        3,123   -> cardinal{ integer: "三千一百二十三"}
+        51234   -> cardinal{ integer: "五万一千二百三十四"}
+        51,234  -> cardinal{ integer: "五万一千二百三十四"}
+        0.125   -> cardinal{ integer: "零点一二五"}
+        self.fst:
+        123     -> cardinal{ integer: "一二三" }
```
[the remark this answers is not in the document]
Comment (mzxcpp) [reply nemo_text_processing/text_normalization/zh/taggers/cardinal.py:40]: Thank you for your advice！The opinion I  handle with number is:①when there is a unit,we read it into cardinal forms,just like "20头牛、125个人、3000元",so the measure part is used to normalize number with unit both in English and Chinese.②When it comes to a single number, just like  "747 or 10086"only, although we prefer reading it to"七百四十七 or 一万零八十六", in real cases,the correct forms may be “波音七四七、编号七四七人员、拨打一二三零六".
To put it simply，the previous work just transport number into cardinal forms and I just found it has a lower accuracy in lots of real cases we tested.Therefore,to handle with more detailed cases,I set it more detailed,just like now.I think we could not deal with all the cases by one form,and the present forms may lead to more accuracy in Chinese Text Normalization.

**Comment by dophist** [comment]
@mzxcpp please give clear signals here once you have addressed all issues from last review, so that we know when to move forward, and everyone won't get distracted by intermediate development activities.

**Comment by mzxcpp** [comment]
Thank you for your detailed review again!I have fixed some naming problems and now make sure the pytest will work.However,with only TN part I have designed,it may be difficult for me to deal with work with  sparrowhawk related.

File: nemo_text_processing/text_normalization/zh/taggers/cardinal.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (yzhang123) [reply nemo_text_processing/text_normalization/zh/taggers/cardinal.py:40]: Thank you @mzxcpp  for the explanation. Very interesting observation. @BuyuanCui, did you see similar things like this and can confirm this?
@mzxcpp is it possible to share the real cases you tested on? that would be highly interesting.

[the remark this answers is not in the document]
Comment (mzxcpp) [reply nemo_text_processing/text_normalization/zh/taggers/cardinal.py:40]: Thank you for taking my idea into consideration!I just base on real cases in Leaderboard.
https://github.com/SpeechColab/Leaderboard
and
https://github.com/dophist/textnorm/blob/main/zh/misc/speechio_0_15.tsv
and I have just gone through some cases and find the following cases:

原知识产权5条热线和平台的整合，实现12315一个号码对外一个平台受理，方便消费者。
针对禁止报出的波音737ng型飞机被发现了机身出现裂缝一事。美国波音公司10月31日宣布。
在5个月内连续发生两次空难后，目前波音公司已暂停向客户交付t37-8及t37-9新飞机。
当天在伊朗首都德黑兰附近机场失事的波音707货机的黑匣子已经找到。
春节期间，各界群众若发现涉黄涉非违法行为，可通过全国扫黄打非办公室举报电话12390。
鲁豫探访奇虎360，带您走进这位互联网传奇老炮周弘毅。
2002年一度领先市场份额的3721逐渐迎来强劲对手。
感谢3158创业信息网的大力支持。
您这个是号是000~008。
身高1米84的克洛泽似乎是在有意避开人们对他只会投球的质疑。
最近呢有个小朋友跟我说，呢他们公司准备要开展996工作制，就是早上9点工作到晚上9点。
1234这4个队员会选择进攻。这样一来先防守他们。
510有点有一点点的偏局红调唉510我试一下吧510应该还可以。我试一下510好不好？
DC37烟熏玫瑰色，所有女生都可以用的颜色。

These cases will go wrong if we set each number into cardinal forms.And it's difficult to try to cover all the special forms with number string.I think it may be easier to cover all the common units rather than all this special forms.

Admittedly,there are cases with wrong answer:

一年每年呢给你免24,000，如果你是非独生子女。
9999995这么大。

In these cases,speaker ignores the unit because in some situation,others can know the correct meaning in Chinese,just like "元" may be left out.
Just my experimental idea and I will follow your ideas whether to set it as cardinal forms or keep the current forms.

**Comment by yzhang123** [comment]
> Thank you for your detailed review again!I have fixed some naming problems and now make sure the pytest will work.However,with only TN part I have designed,it may be difficult for me to deal with work with sparrowhawk related.


Is your TN output non-deterministic? if not, sparrowhawk should run Out of the box

**Comment by mzxcpp** [comment]
> > Thank you for your detailed review again!I have fixed some naming problems and now make sure the pytest will work.However,with only TN part I have designed,it may be difficult for me to deal with work with sparrowhawk related.
> 
> Is your TN output non-deterministic? if not, sparrowhawk should run Out of the box

Yes,the TN output is deterministic and I think it will work sucessfully in this case.However,when i test this,there is an error as "ModuleNotFoundError: No module named 'nemo_text_processing'".Should I install an another version of nemo in my env instead of running the sparrowhawk test in local nemo files?

**Comment by yzhang123** [comment]
> > > Thank you for your detailed review again!I have fixed some naming problems and now make sure the pytest will work.However,with only TN part I have designed,it may be difficult for me to deal with work with sparrowhawk related.
> > 
> > 
> > Is your TN output non-deterministic? if not, sparrowhawk should run Out of the box
> 
> Yes,the TN output is deterministic and I think it will work sucessfully in this case.However,when i test this,there is an error as "ModuleNotFoundError: No module named 'nemo_text_processing'".Should I install an another version of nemo in my env instead of running the sparrowhawk test in local nemo files?

you need to install nemo and nemo_text_processing is installed using https://github.com/NVIDIA/NeMo/blob/main/nemo_text_processing/install_pynini.sh

**Comment by mzxcpp** [comment]
> > > > Thank you for your detailed review again!I have fixed some naming problems and now make sure the pytest will work.However,with only TN part I have designed,it may be difficult for me to deal with work with sparrowhawk related.
> > > 
> > > 
> > > Is your TN output non-deterministic? if not, sparrowhawk should run Out of the box
> > 
> > 
> > Yes,the TN output is deterministic and I think it will work sucessfully in this case.However,when i test this,there is an error as "ModuleNotFoundError: No module named 'nemo_text_processing'".Should I install an another version of nemo in my env instead of running the sparrowhawk test in local nemo files?
> 
> you need to install nemo and nemo_text_processing is installed using https://github.com/NVIDIA/NeMo/blob/main/nemo_text_processing/install_pynini.sh

Thank you!But what can I do if a tagger is not included in sparrowawk?Should I remove them all or keep them without test_sparrowhawk?  I find that if a tagger's form is not included in the semiotic_classes.proto,it will not work successfully.

**Comment by yzhang123** [comment]
> > > > > Thank you for your detailed review again!I have fixed some naming problems and now make sure the pytest will work.However,with only TN part I have designed,it may be difficult for me to deal with work with sparrowhawk related.
> > > > 
> > > > 
> > > > Is your TN output non-deterministic? if not, sparrowhawk should run Out of the box
> > > 
> > > 
> > > Yes,the TN output is deterministic and I think it will work sucessfully in this case.However,when i test this,there is an error as "ModuleNotFoundError: No module named 'nemo_text_processing'".Should I install an another version of nemo in my env instead of running the sparrowhawk test in local nemo files?
> > 
> > 
> > you need to install nemo and nemo_text_processing is installed using https://github.com/NVIDIA/NeMo/blob/main/nemo_text_processing/install_pynini.sh
> 
> Thank you!But what can I do if a tagger is not included in sparrowawk?Should I remove them all or keep them without test_sparrowhawk? I find that if a tagger's form is not included in the semiotic_classes.proto,it will not work successfully. The other part just work successfully like time or currency in this TN.

which classes' tagger is not included? optimally we want to have a  semiotic_classes.proto or expand it so it supports all classes. you can add the pytest still, just don't include them in test_sparrowhawk until we figured out how to expand the .proto file

**Comment by mzxcpp** [comment]
> > > > > > Thank you for your detailed review again!I have fixed some naming problems and now make sure the pytest will work.However,with only TN part I have designed,it may be difficult for me to deal with work with sparrowhawk related.
> > > > > 
> > > > > 
> > > > > Is your TN output non-deterministic? if not, sparrowhawk should run Out of the box
> > > > 
> > > > 
> > > > Yes,the TN output is deterministic and I think it will work sucessfully in this case.However,when i test this,there is an error as "ModuleNotFoundError: No module named 'nemo_text_processing'".Should I install an another version of nemo in my env instead of running the sparrowhawk test in local nemo files?
> > > 
> > > 
> > > you need to install nemo and nemo_text_processing is installed using https://github.com/NVIDIA/NeMo/blob/main/nemo_text_processing/install_pynini.sh
> > 
> > 
> > Thank you!But what can I do if a tagger is not included in sparrowawk?Should I remove them all or keep them without test_sparrowhawk? I find that if a tagger's form is not included in the semiotic_classes.proto,it will not work successfully. The other part just work successfully like time or currency in this TN.
> 
> which classes' tagger is not included? optimally we want to have a semiotic_classes.proto or expand it so it supports all classes. you can add the pytest still, just don't include them in test_sparrowhawk until we figured out how to expand the .proto file

Well,I checked my code again and reorganize it.
I think math in tagger in which I designed detailed graph for cases like "25:26" into"二十五比二十六"，may be we call this "score"
All in all,I think these are just small cases and I have work succeessfully with "pytest" and "sparrowhawk" test.I upload the latest version.
Although there may more detailed cases I don't cover,I think it now can handle with most common cases in Chinese and I hope this can be a base of Chinese TN so that we can add more details to get a perfect TN tool for Chinese in further works.

**Comment by mzxcpp** [comment]
The local setup fix nothing for current code but I can't pass CI test this time.
Is it because I changed something in tools/text_processing_deployment?

**Comment by pengzhendong** [comment]
`他说：“我们已经吃过了！”。` => `他说比“我们已经吃过了!”。`
`这儿有只鸟儿` => `normalize.py", line 363, in _permute raise ValueError()`

**Comment by mzxcpp** [comment]
> `他说：“我们已经吃过了！”。` => `他说比“我们已经吃过了!”。` `这儿有只鸟儿` => `normalize.py", line 363, in _permute raise ValueError()`

Thank you!I have fixed it.

**Comment by XuesongYang** [comment]
Could you please try to fix the alerts as described here: https://lgtm.com/projects/g/NVIDIA/NeMo/rev/pr-f0d82e6d82662b60dc1c597bc403b4ac6dc1615a? Thanks.

**Comment by mzxcpp** [comment]
> Could you please try to fix the alerts as described here: https://lgtm.com/projects/g/NVIDIA/NeMo/rev/pr-f0d82e6d82662b60dc1c597bc403b4ac6dc1615a? Thanks.

Ok,I have deleted the useless lines in the lastest commit

File: nemo_text_processing/text_normalization/zh/graph_utils.py [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+    Args:
+        name: name of grammar class
+        kind: either 'classify' or 'verbalize'
+        deterministic: if True will provide a single transduction option,
+            for False multiple transduction are generated (used for audio-based normalization)
+    """
+
+    def __init__(self, name: str, kind: str, deterministic: bool = True):
+        self.name = name
+        self.kind = str
```
Comment (pengzhendong) [inline nemo_text_processing/text_normalization/zh/graph_utils.py:81]: `self.kind = kind`

**Review by pengzhendong — CHANGES_REQUESTED** [review]

File: nemo_text_processing/text_normalization/zh/graph_utils.py (hunk 14 above)
Comment (pengzhendong) [reply nemo_text_processing/text_normalization/zh/graph_utils.py:81]: There are some useless variables in `graph_utils.py`, such as `HEX`, `NON_BREAKING_SPACE`, `GRAPH`, and so on.

Comment (mzxcpp) [reply nemo_text_processing/text_normalization/zh/graph_utils.py:81]: > `self.kind = kind`

Thank you.I just copy that file from en and I will remove useless things.But self.kind = str is also in en.Is there also a bug in that?

File: nemo_text_processing/text_normalization/zh/taggers/cardinal.py [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+# limitations under the License.
+import pynini
+from nemo_text_processing.text_normalization.zh.graph_utils import GraphFst
+from nemo_text_processing.text_normalization.zh.utils import UNIT_1e01, UNIT_1e02, UNIT_1e03, UNIT_1e04, get_abs_path
+from pynini.lib import pynutil
+
+
+class Cardinal(GraphFst):
+    '''
+        self.graph_cardinal:
```
Comment (pengzhendong) [inline nemo_text_processing/text_normalization/zh/taggers/cardinal.py:22]: `self.graph_cardinal` does not contain tokens. It should be

```
5  -> 五
12 -> 十二
```

**Review by pengzhendong — CHANGES_REQUESTED** [review]

File: nemo_text_processing/text_normalization/zh/graph_utils.py (hunk 14 above)
Comment (pengzhendong) [reply nemo_text_processing/text_normalization/zh/graph_utils.py:81]: I think it is a bug.

File: nemo_text_processing/text_normalization/zh/taggers/cardinal.py [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+                + pynutil.insert(UNIT_1e03)
+                + delete_punct
+                + graph_zero
+                + graph_digit
+                + pynutil.insert(UNIT_1e01)
+                + graph_digit_no_zero
+            )
+            | (graph_digit + pynutil.insert(UNIT_1e03) + delete_punct + graph_zero + graph_no_zero + graph_digit)
+            | (graph_digit + pynutil.insert(UNIT_1e03) + delete_punct + graph_no_zero ** 3)
+        )
```
Comment (pengzhendong) [inline nemo_text_processing/text_normalization/zh/taggers/cardinal.py:85]:
```
|  closure(...)
 |      closure(self, lower)
 |      closure(self, lower, upper)
 |      
 |      Computes concatenative closure.
 |      
 |      This operation destructively converts the FST to its concatenative closure.
 |      If A transduces string x to y with weight w, then the zero-argument form
 |      `A.closure()` mutates A so that it transduces between empty strings with
 |      weight 1, transduces string x to y with weight w, transduces xx to yy with
 |      weight w \otimes w, string xxx to yyy with weight w \otimes w \otimes w
 |      (and so on).
 |      
 |      When called with two optional positive integer arguments, these act as
 |      lower and upper bounds, respectively, for the number of cycles through the
 |      original FST that the mutated FST permits. Therefore, `A.closure(0, 1)`
 |      mutates A so that it permits 0 or 1 cycles; i.e., the mutated A transduces
 |      between empty strings or transduces x to y.
 |      
 |      When called with one optional positive integer argument, this argument
 |      acts as the lower bound, with the upper bound implicitly set to infinity.
 |      Therefore, `A.closure(1)` performs a mutation roughly equivalent to
 |      `A.closure()` except that the former does not transduce between empty
 |      strings.
 |      
 |      The following are the equivalents for the closure-style syntax used in
 |      Perl-style regular expressions:
 |      
 |      Regexp:             This method:            Copy shortcuts:
 |      
 |      /x?/                x.closure(0, 1)         x.ques
 |      /x*/                x.closure()             x.star
 |      /x+/                x.closure(1)            x.plus
 |      /x{N}/              x.closure(N, N)         x ** N
 |      /x{M,N}/            x.closure(M, N)
 |      /x{N,}/             x.closure(N)
 |      /x{,N}/             x.closure(0, N)
```

Please try to use `x.ques` to avoid the redundant code.

**Review by pengzhendong — CHANGES_REQUESTED** [review]

File: nemo_text_processing/text_normalization/zh/taggers/cardinal.py (hunk 16 above)
Comment (pengzhendong) [inline nemo_text_processing/text_normalization/zh/taggers/cardinal.py:85]:
```
|  Regexp:  This method:     Copy shortcuts:
|  /x?/     x.closure(0, 1)  x.ques
```

I think it's better to use `delete_punct.ques` to avoid the redundant code.

**Review by pengzhendong — CHANGES_REQUESTED** [review]

File: nemo_text_processing/text_normalization/zh/taggers/char.py [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+
+class Char(GraphFst):
+    '''
+        你 -> tokens { char: "你" }
+    '''
+
+    def __init__(self, deterministic: bool = True, lm: bool = False):
+        super().__init__(name="char", kind="classify", deterministic=deterministic)
+
+        graph = pynutil.insert("name: \"") + NEMO_CHAR + pynutil.insert("\"")
```
Comment (pengzhendong) [inline nemo_text_processing/text_normalization/zh/taggers/char.py:26]: `pynutil.insert("char: \"")`

**Review by pengzhendong — CHANGES_REQUESTED** [review]

File: nemo_text_processing/text_normalization/zh/taggers/char.py (hunk 17 above)
Comment (mzxcpp) [reply nemo_text_processing/text_normalization/zh/taggers/char.py:26]: It used to be "char" and I followed the previous suggestions from ekmb to change it into name as the basic char.Should I change it to char again?

Comment (pengzhendong) [reply nemo_text_processing/text_normalization/zh/taggers/char.py:26]: OK, may be you should chang this line, too.

```
你 -> tokens { char: "你" }
```

Comment (mzxcpp) [reply nemo_text_processing/text_normalization/zh/taggers/char.py:26]: Ok.I have solved the above suggestions.

**Comment by yzhang123** [comment]
@mzxcpp could you please remove the unrelated files from your PR:

image

After that, i think we can merge it for now. Verified that pytests and SH tests are passing.

File: nemo_text_processing/text_normalization/zh/graph_utils.py (hunk 14 above)
Comment (yzhang123) [reply nemo_text_processing/text_normalization/zh/graph_utils.py]: yes it is, thanks for catching it @pengzhendong , @mzxcpp could you please fix it in this file and all other files?

**Review by yzhang123 — CHANGES_REQUESTED** [review]
